quantum/process_keycode/process_autocorrect.c

```diff
@@ -25,65 +22,65 @@
  * @return true Continue processing keycodes, and send to host
  * @return false Stop processing keycodes, and don't send to host
  */
-bool process_autocorrection(uint16_t keycode, keyrecord_t* record) {
-    static uint8_t typo_buffer[AUTOCORRECTION_MAX_LENGTH] = {KC_SPC};
-    static uint8_t typo_buffer_size                       = 1;
 
-    if (keycode == AUTO_CTN) {
+bool process_autocorrect(uint16_t keycode, keyrecord_t* record) {
+    static uint8_t typo_buffer[AUTOCORRECT_MAX_LENGTH] = {KC_SPC};
+    static uint8_t typo_buffer_size                    = 1;
+
+    if (keycode == AUTOCRT) {
         if (record->event.pressed) {
             typo_buffer_size = 0;
-            userspace_config.autocorrection ^= 1;
-            eeconfig_update_user(userspace_config.raw);
+            keymap_config.autocorrect_enable ^= 1;
+            eeconfig_update_keymap(keymap_config.raw);
         }
         return false;
     }
 
-    if (!userspace_config.autocorrection) {
+    if (!keymap_config.autocorrect_enable) {
         typo_buffer_size = 0;
         return true;
     }
 
+    if (!record->event.pressed) {
+        return true;
+    }
+
     switch (keycode) {
         case KC_LSFT:
         case KC_RSFT:
             return true;
-#    ifndef NO_ACTION_TAPPING
+#ifndef NO_ACTION_TAPPING
         case QK_MOD_TAP ... QK_MOD_TAP_MAX:
-            if (((keycode >> 8) & 0xF) == MOD_LSFT) {
-                return true;
-            }
-#        ifndef NO_ACTION_LAYER
+#    ifndef NO_ACTION_LAYER
         case QK_LAYER_TAP ... QK_LAYER_TAP_MAX:
-#        endif
-            if (record->event.pressed || !record->tap.count) {
-                return true;
+#    endif
+            if (!record->tap.count) {
+                return true;  // Early return if tap-hold key is not being tapped.
             }
             keycode &= 0xFF;
             break;
-#    endif
-#    ifdef SWAP_HANDS_ENABLE
+#endif
+#ifdef SWAP_HANDS_ENABLE
         case QK_SWAP_HANDS ... QK_SWAP_HANDS_MAX:
-            if (keycode >= 0x56F0 || record->event.pressed || !record->tap.count) {
+            if (keycode >= 0x56F0 || !record->tap.count) {
                 return true;
             }
             keycode &= 0xFF;
             break;
-#    endif
-#    ifndef NO_ACTION_ONESHOT
+#endif
+#ifndef NO_ACTION_ONESHOT
         case QK_ONE_SHOT_MOD ... QK_ONE_SHOT_MOD_MAX:
             if ((keycode & 0xF) == MOD_LSFT) {
                 return true;
             }
-#    endif
+            // fall through intentionally
+#endif
         default:
-            // Disable autocorrection while a mod other than shift is active.
+            // Disable autocorrect while a mod other than shift is active.
             if (((get_mods() | get_oneshot_mods()) & ~MOD_MASK_SHIFT) != 0) {
                 typo_buffer_size = 0;
                 return true;
             }
-            if (!record->event.pressed) {
-                return true;
-            }
     }
 
```

Comment:
Here is a suggestion for a small, but I think worthwhile, improvement to autocorrect.

Currently, autocorrect handles typos containing the chars a-z and the special "word break" char (:). I realized today that autocorrect could be extended to handle typos containing the ' single quote char, which would be useful for autocorrection entries involving apostrophes (example: `dosen't -> doesn't`). This is doable with a few small changes. 

Here in the firmware, modify this code with a branch for KC_QUOT. A detail is that double quote " (shifted ') should rather be a word break.

```c
if (keycode == KC_QUOT) {
    // Treat " (shifted ') as a word boundary.
    if ((mods & MOD_MASK_SHIFT) != 0) { keycode = KC_SPC; }
} else if (!(KC_A <= keycode && keycode <= KC_Z)) {
   Same as before ...
```

In the Python code, define

```py
KC_QUOT = 0x34

TYPO_CHARS = dict(
  [
    ("'", KC_QUOT),
    (':', KC_SPC),  # "Word break" character.
  ] +
  # Characters a-z.
  [(chr(c), c + KC_A - ord('a')) for c in range(ord('a'), ord('z') + 1)]
)
```

Then in `parse_file()`, make the change

```py
# if not (all([ord('a') <= ord(c) <= ord('z') or c == ':' for c in typo])):
if not(all([c in TYPO_CHARS for c in typo])):
```

And in `serialize()`, remove the `kc_code()` function and replace it with lookup in `TYPO_CHARS`:

```py
def serialize(e):
    ...
    # return list(map(kc_code, e['chars'])) + [0]
    return [TYPO_CHARS[c] for c in e['chars']] + [0]
    ...
    # data += [kc_code(c) | (0 if data else 64)] + encode_link(link)
    data += [TYPO_CHARS[c] | (0 if data else 64)] + encode_link(link)
```

I just checked in this changes to my repo at
https://github.com/getreuer/qmk-keymap/commit/d8f6562744cc636cea9b6c941c68e2da864fc6cb

Reply:
That's definitely a great change!